src/condor_schedd.V6/schedd_negotiate.h

```diff
@@ -23,50 +23,75 @@
  *
- * The auto cluster ids are only assumed to be internally consistent
- * at the time the snapshot of the job queue was taken.  They are not used
- * to reference any external data structure.
  */
 class ResourceRequestCluster {
- public:
-	ResourceRequestCluster(int auto_cluster_id);
-
-		// appends a job to this cluster
-	void addJob(PROC_ID job_id);
-
-		// returns true if a job was found; o.w. false
-	bool popJob(PROC_ID &job_id);
-
-		// returns number of jobs in this cluster
-	size_t size() { return m_job_ids.size(); }
-
-		// returns the auto cluster id for this cluster
-	int getAutoClusterId() const { return m_auto_cluster_id; }
- private:
-
-	int m_auto_cluster_id;
-	std::list<PROC_ID> m_job_ids;
+public:
+	ResourceRequestCluster(int id=0, int capacity=0) : m_autocluster_id(id) {
+		if (capacity) m_jobs.reserve(capacity);
+	}
+	void addJob(PROC_ID jid) { m_jobs.push_back(jid); }
+
+	// returns number of jobs in this cluster
+	size_t size() { return m_jobs.size(); }
+
+	// returns the auto cluster id for this cluster
+	int getAutoClusterId() const { return m_autocluster_id; }
+
+	using container = std::vector<PROC_ID>;
+	using iterator = container::iterator;
+	using const_iterator = container::const_iterator;
+	iterator begin() { return m_jobs.begin(); }
+	const_iterator begin() const { return m_jobs.cbegin(); }
+	iterator end() { return m_jobs.end(); }
+	const_iterator end() const { return m_jobs.cend(); }
+
+protected:
+	std::vector<PROC_ID> m_jobs;
+	int                  m_autocluster_id{0};
 };
 
 /*
- * ResourceRequestList
- *
- * Contains an ordered list of job auto clusters for matchmaking.  All
- * jobs in the list are owned by the same user (or accounting group).
- * 
- * The ResourceRequestClusters in this list are deleted when the list
- * is deleted.
- */
-class ResourceRequestList: public std::list<ResourceRequestCluster *> {
- public:
-	~ResourceRequestList();
+* ResourceRequestList
+*
+* Contains an ordered list of job auto clusters for matchmaking.  All
+* jobs in the list are owned by the same user (or accounting group).
+* 
+* The list is owned by an instance of ScheddNegotiate and consumed (deleted) by it
+* as the conversation proceeds. Currently ResourceRequestClusters in this list
+* and are also deleted during negotiation, but planned future changes will
+* Make the resource request clusters persist, but list will not.
+*/
+class ResourceRequestList {
+public:
+	void add(int id, PROC_ID & jid) {
+		if ( ! items.empty() && items.back().getAutoClusterId() == id) {
+			items.back().addJob(jid);
+		} else {
+			items.emplace_back(id).addJob(jid);
+		}
+	}
+	int currentId() {
+		if (items.empty()) return -1;
+		return items.back().getAutoClusterId();
+	}
+	bool empty() { return items.empty(); }
+	size_t size() { return items.size(); }
+	ResourceRequestCluster & front() { return items.front(); }
+	void pop_front() { return items.pop_front(); }
```

Comment:
The return here implies that pop_front() could return a value. We should just remove it.

Reply:
I was a bit surprised that c++ considered it valid, but it does..